ambry-store/src/main/java/com.github.ambry.store/HardDeleter.java

```diff
@@ -111,17 +113,30 @@ public void run() {
     try {
       while (running.get()) {
         try {
-          if (!hardDelete()) {
-            isCaughtUp = true;
+          if (paused.get()) {
```

Comment:
two questions regarding this logic.

1. Why isn't this

```
synchronized(lock) {
  while(paused) { // not sure why this has to be a "while" and not an "if"
    wait();
  }
  // hard delete code
}
```
Why are you taking the lock twice in a row? Code looks a bit more complicated than it has to be.

2. Why are you using the same object (in essence the same condition) to notify()  both the requirement to pause/unpause and the fact that pause has completed? I don't know if there is a correctness issue but it makes the code confusing. Why not use a `ReentrantLock` with two `Condition` instances attached to it?